// @trezor/rollout/lib/utils/version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNewerOrEqual = exports.isEqual = exports.isNewer = exports.toString = exports.parse = void 0;
exports.parse = function (versionArr) {
    return {
        major: versionArr[0],
        minor: versionArr[1],
        patch: versionArr[2],
    };
};
exports.toString = function (arr) { return arr[0] + "." + arr[1] + "." + arr[2]; };
exports.isNewer = function (versionX, versionY) {
    var parsedX = exports.parse(versionX);
    var parsedY = exports.parse(versionY);
    if (parsedX.major - parsedY.major !== 0) {
        return parsedX.major > parsedY.major;
    }
    if (parsedX.minor - parsedY.minor !== 0) {
        return parsedX.minor > parsedY.minor;
    }
    if (parsedX.patch - parsedY.patch !== 0) {
        return parsedX.patch > parsedY.patch;
    }
    return false;
};
exports.isEqual = function (versionX, versionY) {
    return exports.toString(versionX) === exports.toString(versionY);
};
exports.isNewerOrEqual = function (versionX, versionY) {
    return exports.isNewer(versionX, versionY) || exports.isEqual(versionX, versionY);
};
